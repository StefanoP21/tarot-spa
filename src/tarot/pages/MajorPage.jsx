import { TarotList } from '../components/TarotList';

export const MajorPage = () => {
  return <TarotList type="Major Arcana" />;
};
