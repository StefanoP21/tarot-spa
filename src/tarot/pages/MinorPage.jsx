import { TarotList } from '../components/TarotList';

export const MinorPage = () => {
  return <TarotList type="Minor Arcana" />;
};
