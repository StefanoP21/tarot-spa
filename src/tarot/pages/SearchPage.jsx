import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';

import { getTarotByName } from '../helpers/getTarotByName';
import { useForm } from '../../hooks/useForm';
import { TarotCard, TarotMessage } from '../components';

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = '' } = queryString.parse(location.search);

  const tarot = getTarotByName(q);

  const showSearch = q.length === 0;
  const showError = q.length > 0 && tarot.length === 0;

  const { searchText, onInputChange } = useForm({
    searchText: q,
  });

  const onSearchSubmit = (e) => {
    e.preventDefault();

    navigate(`?q=${searchText}`);
  };

  return (
    <div className="bg-gray-700">
      <div className="mx-auto max-w-xl px-4 py-10 sm:px-6 sm:py-10 lg:max-w-7xl lg:px-8">
        <form onSubmit={onSearchSubmit}>
          <div className="sm:col-span-4">
            <label
              htmlFor="searchText"
              className="block text-sm font-bold leading-6 text-white"
            >
              Card Name
            </label>

            <div className="mt-2">
              <div className="flex rounded-md shadow-sm ring-1 ring-inset ring-gray-300 focus-within:ring-2 focus-within:ring-inset focus-within:ring-purple-600 sm:max-w-md">
                <input
                  type="text"
                  name="searchText"
                  id="searchText"
                  value={searchText}
                  onChange={onInputChange}
                  className="block flex-1 border-0 bg-transparent py-1.5 pl-2 text-gray-200 placeholder:text-gray-400 focus:ring-0 sm:text-sm sm:leading-6"
                  placeholder="The Fool"
                />
              </div>
            </div>

            <div>
              <button
                type="submit"
                className="mt-4 bg-purple-600 border border-transparent rounded-md py-2 px-4 inline-flex justify-center text-sm font-bold text-white hover:bg-purple-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-purple-500"
              >
                Search
              </button>
            </div>
          </div>
        </form>

        {showSearch && (
          <TarotMessage
            title="Search for a Tarot Card"
            message='Use the search bar above to find a tarot card by name. For example,
            try searching for "The Fool" or "The Magician".'
            color="bg-purple-900"
          />
        )}

        {showError && (
          <TarotMessage
            title="No Results Found"
            message="We couldn't find any tarot cards that matched your search. Please try again."
            color="bg-red-900"
          />
        )}

        <div className="mt-10 mb-20 grid grid-cols-1 gap-x-6 gap-y-10 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 xl:gap-x-8">
          {tarot.map((card) => (
            <TarotCard key={card.id} {...card} />
          ))}
        </div>
      </div>
    </div>
  );
};
