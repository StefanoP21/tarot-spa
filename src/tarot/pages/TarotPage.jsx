import { useMemo } from 'react';
import { Link, Navigate, useNavigate, useParams } from 'react-router-dom';
import { getTarotbyId } from '../helpers/getTarotbyId';

export const TarotPage = () => {
  const { id } = useParams();
  const navigate = useNavigate();

  const tarot = useMemo(() => getTarotbyId(id), [id]);

  const onNavigateBack = () => {
    navigate(-1);
  };

  if (!tarot) {
    return <Navigate to="/major" />;
  }

  return (
    <div className="bg-gray-700">
      <div className="pt-6">
        <nav aria-label="Breadcrumb">
          <ol
            role="list"
            className="mx-auto flex max-w-2xl items-center space-x-2 px-4 sm:px-6 lg:max-w-7xl lg:px-8"
          >
            <li>
              <div className="flex items-center">
                <Link
                  to={`/${tarot.path}`}
                  className="mr-2 text-sm font-medium text-white"
                >
                  {tarot.type}
                </Link>
                <svg
                  width={16}
                  height={20}
                  viewBox="0 0 16 20"
                  fill="currentColor"
                  aria-hidden="true"
                  className="h-5 w-4 text-gray-300"
                >
                  <path d="M5.697 4.34L8.98 16.532h1.327L7.025 4.341H5.697z" />
                </svg>
              </div>
            </li>

            <li className="text-sm">
              <a
                href={tarot.id}
                aria-current="page"
                className="font-medium text-gray-400 hover:text-gray-200"
              >
                {tarot.name}
              </a>
            </li>
          </ol>
        </nav>

        {/* Image */}
        <div className="mx-auto mt-6 max-w-2xl sm:px-6 lg:grid lg:max-w-7xl lg:grid-cols-3 lg:gap-x-8 lg:px-8">
          <div className="aspect-h-5 aspect-w-4 lg:aspect-h-4 lg:aspect-w-3 sm:overflow-hidden sm:rounded-lg">
            <img
              src={`/public/assets/tarot/${tarot.id}.webp`}
              alt={tarot.name}
              className="h-full w-full object-cover object-center"
            />
          </div>
        </div>

        {/* Card info */}
        <div className="mx-auto max-w-2xl px-4 pb-16 pt-10 sm:px-6 lg:grid lg:max-w-7xl lg:grid-cols-3 lg:grid-rows-[auto,auto,1fr] lg:gap-x-8 lg:px-8 lg:pb-24 lg:pt-16">
          <div className="lg:col-span-2 lg:border-r lg:border-gray-200 lg:pr-8">
            <h1 className="text-2xl font-bold tracking-tight text-white sm:text-3xl">
              {tarot.name}
            </h1>
          </div>

          {/* Location */}
          <div className="mt-4 lg:row-span-3 lg:mt-0">
            <h2 className="sr-only">Product information</h2>
            <p className="text-xl tracking-tight text-gray-200">
              {tarot.location}
            </p>

            <button
              className="mt-10 flex w-full items-center justify-center rounded-md border border-transparent bg-purple-700 px-8 py-3 text-base font-medium text-white hover:bg-purple-800 focus:ring-4 focus:outline-none focus:ring-purple-600 focus:ring-offset-2"
              onClick={onNavigateBack}
            >
              Back
            </button>
          </div>

          <div className="py-10 lg:col-span-2 lg:col-start-1 lg:border-r lg:border-gray-200 lg:pb-16 lg:pr-8 lg:pt-6">
            {/* Journal and Description */}
            <div>
              <h2 className="text-md font-bold text-gray-100">Journal</h2>
              <div className="mt-4 space-y-6">
                <p className="text-base text-gray-300">{tarot.journal}</p>
              </div>
            </div>

            <div className="mt-10">
              <h2 className="text-sm font-bold text-gray-200">Description</h2>
              <div className="mt-4 space-y-6">
                <p className="text-sm text-gray-400">{tarot.description}</p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};
