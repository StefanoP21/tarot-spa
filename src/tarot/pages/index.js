export * from './MajorPage';
export * from './MinorPage';
export * from './TarotPage';
export * from './SearchPage';
