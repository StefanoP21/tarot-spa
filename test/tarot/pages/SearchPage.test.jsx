import { fireEvent, render, screen } from '@testing-library/react';
import { SearchPage } from '../../../src/tarot/pages/SearchPage';
import { MemoryRouter } from 'react-router-dom';

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),

  useNavigate: () => mockedUsedNavigate,
}));

describe('Pruebas en <SearchPage />', () => {
  beforeEach(() => jest.clearAllMocks());

  test('debe de mostrarse correctamente con valores por defecto', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );

    // screen.debug();
    expect(container).toMatchSnapshot();
  });

  test('debe de mostrar la imagen de "fool" y el input con el valor del queryString', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=fool']}>
        <SearchPage />
      </MemoryRouter>
    );

    // screen.debug();
    const img = screen.getByRole('img');
    expect(img.src).toContain('the-fool.webp');

    const input = screen.getByRole('textbox');
    expect(input.value).toBe('fool');
  });

  test('debe de mostrar el componente con el mensaje de busqueda', () => {
    render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );

    // screen.debug();
    const message = screen.getByText('Search for a Tarot Card');
    expect(message).toBeTruthy();
  });

  test('debe de mostrar el componente con el mensaje de error', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=xyz']}>
        <SearchPage />
      </MemoryRouter>
    );

    // screen.debug();
    const message = screen.getByRole('heading', { name: 'No Results Found' });
    expect(message).toBeTruthy();
  });

  test('debe de llamar el navigate a la pantalla nueva', () => {
    const inputValue = 'fool';

    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage />
      </MemoryRouter>
    );

    const input = screen.getByRole('textbox');
    fireEvent.change(input, {
      target: { name: 'searchText', value: inputValue },
    });

    const searchButton = screen.getByRole('button');
    fireEvent.click(searchButton);

    expect(mockedUsedNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);
  });
});
