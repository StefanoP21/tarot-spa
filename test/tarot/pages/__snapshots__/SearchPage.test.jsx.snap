// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <div
    class="bg-gray-700"
  >
    <div
      class="mx-auto max-w-xl px-4 py-10 sm:px-6 sm:py-10 lg:max-w-7xl lg:px-8"
    >
      <form>
        <div
          class="sm:col-span-4"
        >
          <label
            class="block text-sm font-bold leading-6 text-white"
            for="searchText"
          >
            Card Name
          </label>
          <div
            class="mt-2"
          >
            <div
              class="flex rounded-md shadow-sm ring-1 ring-inset ring-gray-300 focus-within:ring-2 focus-within:ring-inset focus-within:ring-purple-600 sm:max-w-md"
            >
              <input
                class="block flex-1 border-0 bg-transparent py-1.5 pl-2 text-gray-200 placeholder:text-gray-400 focus:ring-0 sm:text-sm sm:leading-6"
                id="searchText"
                name="searchText"
                placeholder="The Fool"
                type="text"
                value=""
              />
            </div>
          </div>
          <div>
            <button
              class="mt-4 bg-purple-600 border border-transparent rounded-md py-2 px-4 inline-flex justify-center text-sm font-bold text-white hover:bg-purple-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-purple-500"
              type="submit"
            >
              Search
            </button>
          </div>
        </div>
      </form>
      <div
        class="mt-10"
      >
        <div
          class="bg-purple-900 px-4 py-5 sm:px-6 sm:py-6 sm:rounded-lg sm:shadow-xl"
        >
          <h3
            class="text-xl font-semibold leading-6 text-white"
          >
            Search for a Tarot Card
          </h3>
          <div
            class="mt-2 text-md text-gray-100"
          >
            <p>
              Use the search bar above to find a tarot card by name. For example,
 try searching for "The Fool" or "The Magician".
            </p>
          </div>
        </div>
      </div>
      <div
        class="mt-10 mb-20 grid grid-cols-1 gap-x-6 gap-y-10 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 xl:gap-x-8"
      />
    </div>
  </div>
</div>
`;
